### PR TITLE
Add dummy data

### PR DESCRIPTION
**Pull request by msimionati**
Dummy services for billing and business partner controllers were added to avoid VPN connection errors for qa and int environments

### CODE REVIEW

File: Doppler.Sap/Services/BusinessPartnerServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+using Doppler.Sap.Models;
```
Comment (nreal) [inline Doppler.Sap/Services/BusinessPartnerServiceCollectionExtensions.cs:1]: This file should be remove

File: Doppler.Sap/Services/BillingServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
+using Doppler.Sap.Models;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Doppler.Sap.Services
+{
+    public static class ServicesCollectionExtensions
```
Comment (andresmoschini) [inline Doppler.Sap/Services/BillingServiceCollectionExtensions.cs:7]: The convention to create extensions to `IServiceCollection` is place them into the namespace `Microsoft.Extensions.DependencyInjection`:

> [...] We recommended that apps follow this naming convention. Place extension methods in the `Microsoft.Extensions.DependencyInjection` namespace to encapsulate groups of service registrations.
> - [Register groups of services with extension methods](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-3.1#register-groups-of-services-with-extension-methods)

File: Doppler.Sap/Services/BillingServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    {
+        public static void AddServices(this IServiceCollection services, IConfiguration configuration)
```
Comment (andresmoschini) [inline Doppler.Sap/Services/BillingServiceCollectionExtensions.cs:9]: Is it possible to use a more descriptive name?

File: Doppler.Sap/Services/BillingServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        {
+            services.AddSingleton<DummyBillingService>();
+            services.AddTransient<BillingService>();
+            services.AddSingleton<DummyBusinessPartnerService>();
+            services.AddTransient<BusinessPartnerService>();
+
+            var section = configuration.GetSection("SapConfig");
+            var sapSettings = new SapConfig();
+            section.Bind(sapSettings);
```
Comment (andresmoschini) [inline Doppler.Sap/Services/BillingServiceCollectionExtensions.cs:18]: Was not this configuration section already binded using the `IOptions<>` pattern?

File: Doppler.Sap/Services/BillingServiceCollectionExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+            services.AddSingleton<IBillingService>(provider =>
+            {
+                if (!section.Exists() || sapSettings.UseDummyData)
+                    return provider.GetRequiredService<DummyBillingService>();
+
+                return provider.GetRequiredService<BillingService>();
```
Comment (andresmoschini) [inline Doppler.Sap/Services/BillingServiceCollectionExtensions.cs:25]: If the configuration has been read before, we could move the condition out of this delegate, what is more, we could register only one service and all of this is not required.

What do you think about using the same pattern than in Billing API and postpone the reading of the configuration to when the service is resolved?

File: Doppler.Sap/Services/BillingServiceCollectionExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+            services.AddSingleton<DummyBillingService>();
+            services.AddTransient<BillingService>();
+            services.AddSingleton<DummyBusinessPartnerService>();
+            services.AddTransient<BusinessPartnerService>();
+
+            var section = configuration.GetSection("SapConfig");
+            var sapSettings = new SapConfig();
+            section.Bind(sapSettings);
+
+            services.AddSingleton<IBillingService>(provider =>
```
Comment (andresmoschini) [inline Doppler.Sap/Services/BillingServiceCollectionExtensions.cs:20]: It was transient before, are you sure about changing the scope?

File: Doppler.Sap/Services/BillingServiceCollectionExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+                return provider.GetRequiredService<BillingService>();
+            });
+
+            services.AddSingleton<IBusinessPartnerService>(provider =>
+            {
+                if (!section.Exists() || sapSettings.UseDummyData)
+                    return provider.GetRequiredService<DummyBusinessPartnerService>();
+
+                return provider.GetRequiredService<BusinessPartnerService>();
+            });
```
Comment (andresmoschini) [inline Doppler.Sap/Services/BillingServiceCollectionExtensions.cs:34]: Please take into account my comments related to `IBillingService` for this one too.

File: Doppler.Sap/Services/BillingServiceCollectionExtensions.cs (hunk 3 above)
Comment (msimionati) [reply Doppler.Sap/Services/BillingServiceCollectionExtensions.cs:9]: For the class or to the method or both?

Comment (andresmoschini) [reply Doppler.Sap/Services/BillingServiceCollectionExtensions.cs:9]: The method, for example:

```suggestion
        public static void AddSapServices(this IServiceCollection services, IConfiguration configuration)
```

File: Doppler.Sap/Services/BillingServiceCollectionExtensions.cs (hunk 5 above)
Comment (msimionati) [reply Doppler.Sap/Services/BillingServiceCollectionExtensions.cs:25]: ok, I am analyzing it and seeing how it was implemented in the billing api

File: Doppler.Sap/Services/ConfigureSapProviderOptions.cs [hunk 8]
```diff
@@ -0,0 +1,22 @@
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Options;
+
+namespace Doppler.Sap.Services
+{
+    public class ConfigureSapProviderOptions : IConfigureOptions<SapProviderOptions>
```
Comment (andresmoschini) [inline Doppler.Sap/Services/ConfigureSapProviderOptions.cs:6]: Me parece que no es necesario crear un `IConfigureOptions` para esto, creo simplemente se puede registar en `Startup` de la siguiente forma:

```csharp
public void ConfigureServices(IServiceCollection services)
{
    // . . .
    services.Configure<SapProviderOptions>(Configuration.GetSection("SapConfig"));
    // . . .
```

File: Doppler.Sap/appsettings.json [hunk 9]
```diff
@@ -13,7 +13,7 @@
   "AllowedHosts": "*",
   "PublicKeysFolder": "public-keys",
   "SapConfig": {
-    "CompanyDB": "MK_ARG_TEST",
+    "CompanyDB": "MSARGPROD",
```
Comment (andresmoschini) [inline Doppler.Sap/appsettings.json:16]: Este cambio parece correcto, pero fuera del scope de este PR, no?

File: Doppler.Sap/appsettings.qa.json [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+  "SapConfig": {
+    "UseDummyData": true,
+    "CompanyDB": "MK_ARG_TEST",
+    "UserName": "interfaz",
+    "Password": "[SECRET_KEY]",
+    "BaseServerUrl": "https://172.25.16.86:50000/b1s/v1/",
+    "SlackAlertUrl": "[SECRET_KEY]",
+    "MaxAmountAllowedAccounts": 10,
+    "SessionTimeoutPadding": 2
+  }
```
Comment (andresmoschini) [inline Doppler.Sap/appsettings.qa.json:24]: `suggestion
  "SapConfig": {
    "UseDummyData": true,
  }
```

 se me escapa algo?

File: Doppler.Sap/Services/ConfigureSapProviderOptions.cs (hunk 8 above)
Comment (andresmoschini) [reply Doppler.Sap/Services/ConfigureSapProviderOptions.cs:6]: De hecho, veo que ya está:

image

¿No será que faltaba esto (no verifiqué)?

```csharp
public void ConfigureServices(IServiceCollection services)
{
    // . . .
    services.AddOptions();
```

File: Doppler.Sap/appsettings.json (hunk 9 above)
Comment (msimionati) [reply Doppler.Sap/appsettings.json:16]: Las key el archivo de settings ya las tenia, no sabría decirte.

Comment (andresmoschini) [reply Doppler.Sap/appsettings.json:16]: En el diff se ve como que agregaras todo

File: Doppler.Sap/Services/BillingServiceCollectionExtensions.cs (hunk 7 above)
Comment (msimionati) [reply Doppler.Sap/Services/BillingServiceCollectionExtensions.cs]: Ok, done

File: Doppler.Sap/appsettings.json (hunk 9 above)
Comment (andresmoschini) [reply Doppler.Sap/appsettings.json:16]: @nreal es correcto hacer esto?

Comment (nreal) [reply Doppler.Sap/appsettings.json:16]: estas variables se estan manejando en el servicio de sap de swarm, es correcto pero habra quedado de una prueba que hicimos, lo reviso

Comment (andresmoschini) [reply Doppler.Sap/appsettings.json:16]: Gracias, ahí lo charlamos, vamos a separar esto junto con otro cambio para simplificar las pruebas locales. @msimionati, luego lo hacemos juntos.

**Comment by nreal** [comment]
👍

**Pull request merged**